### PR TITLE
Invariant syntax

### PR DESCRIPTION
**Pull request by kitlangton**
Building off of #70, I've combined the type params. Just got everything passing. Next step would be currying certain functions to improve inference, making sure all calls to `any`/`unit` do the right thing.

### CODE REVIEW

**Comment by vigoo** [comment]
I also wanted to give this a try today :) 

I think all the extension methods for `Syntax` can be moved to the class itself now.

**Comment by kitlangton** [comment]
Shall we merge this and then experiment further with smaller refactorings? Or want me to keep playing on this one? :)

File: zio-parser/shared/src/test/scala/zio/parser/PrinterSpec.scala [hunk 1]
```diff
@@ -24,13 +24,13 @@ object PrinterSpec extends ZIOSpecDefault {
...
         ),
         printerTest("transform", Syntax.anyChar.transform(_.toInt, (v: Int) => v.toChar), 66)(isRight(equalTo("B"))),
-        printerTest(
-          "transformEither, failing",
-          Syntax.anyChar.transformEither(_ => Left("bad"), (_: Int) => Left("bad")),
-          100
-        )(
-          isLeft(equalTo("bad"))
-        ),
+//        printerTest(
```
Comment (vigoo) [inline zio-parser/shared/src/test/scala/zio/parser/PrinterSpec.scala:27]: What's up with this? :)

File: zio-parser/shared/src/main/scala/zio/parser/Syntax.scala [hunk 2]
```diff
@@ -431,49 +409,49 @@ class Syntax[+Err, -In, +Out, -Value, +Result] private (
...
     new Syntax(parser, printer)
 
   /** Syntax that does not parse or print anything but succeeds with 'value' */
-  def succeed[Result](value: Result): Syntax[Nothing, Any, Nothing, Any, Result] =
+  def succeed[Value](value: Value): Syntax[Nothing, Any, Nothing, Value] =
     new Syntax(Parser.Succeed(value), Printer.Succeed(value))
 
   /** Syntax that does not pares or print anything but fails with 'failure' */
-  def fail[Err](failure: Err): Syntax[Err, Any, Nothing, Any, Nothing] =
+  def fail[Err](failure: Err): Syntax[Err, Any, Nothing, Nothing] = // TODO: DOUBLE CHECK
```
Comment (vigoo) [inline zio-parser/shared/src/main/scala/zio/parser/Syntax.scala:431]: Double check what?

**Comment by vigoo** [comment]
> Shall we merge this and then experiment further with smaller refactorings? Or want me to keep playing on this one? :)

We should merge it I just noticed a commented out a test. Can we reenable that?

File: zio-parser/shared/src/test/scala/zio/parser/PrinterSpec.scala (hunk 1 above)
Comment (kitlangton) [reply zio-parser/shared/src/test/scala/zio/parser/PrinterSpec.scala]: Whoops. Added back.

File: zio-parser/shared/src/main/scala/zio/parser/Syntax.scala (hunk 2 above)
Comment (kitlangton) [reply zio-parser/shared/src/main/scala/zio/parser/Syntax.scala:431]: Ha. That I wasn't doing anything crazy

Comment (kitlangton) [reply zio-parser/shared/src/main/scala/zio/parser/Syntax.scala:431]: But it all worked.

**Comment by kitlangton** [comment]
Reenabled and deleted the comment :D

**Review by vigoo — APPROVED** [review]

**Pull request merged**